include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -927,47 +1156,211 @@ struct __brick_includes
+            __rng_left_vector, __rng);
+        oneapi::dpl::__par_backend_hetero::__vector_store<__base_t::__preferred_vector_size>{__n}(
+            std::true_type{}, __left_start_idx,
+            oneapi::dpl::__par_backend_hetero::__lazy_store_transform_op<oneapi::dpl::__internal::__pstl_assign>{},
+            __rng_right_vector, __rng);
+        // 4. Call destructors of temporary storage
+        oneapi::dpl::__par_backend_hetero::__vector_walk<__base_t::__preferred_vector_size>{__n}(
+            std::true_type{}, 0, oneapi::dpl::__internal::__lazy_ctor_storage<_ValueType>::__get_callable_deleter(),
+            __rng_left_vector);
+        oneapi::dpl::__par_backend_hetero::__vector_walk<__base_t::__preferred_vector_size>{__n}(
+            std::true_type{}, 0, oneapi::dpl::__internal::__lazy_ctor_storage<_ValueType>::__get_callable_deleter(),
+            __rng_right_vector);
+    }
+    template <typename _IsFull, typename _Idx>
+    void
+    __scalar_path_impl(_IsFull, const _Idx __idx, _Range __rng) const
     {
         using ::std::swap;
-        swap(__acc[__idx], __acc[__size - __idx - 1]);
+        swap(__rng[__idx], __rng[__size - __idx - 1]);
+    }
+    template <typename _IsFull, typename _Idx>
+    void
+    operator()(_IsFull __is_full, const _Idx __idx, _Range __rng) const
+    {
+        if constexpr (__base_t::__can_vectorize)
+            __vector_path_impl(__is_full, __idx, __rng);
+        else
+            __scalar_path_impl(__is_full, __idx, __rng);
     }
 };
 
 //------------------------------------------------------------------------
 // reverse_copy
 //------------------------------------------------------------------------
-template <typename _Size>
-struct __reverse_copy
+template <typename _Size, typename _Range1, typename _Range2>
+struct __reverse_copy : public walk_vector_or_scalar_base<_Range1, _Range2>
 {
+  private:
+    using __base_t = walk_vector_or_scalar_base<_Range1, _Range2>;
+    using _ValueType = oneapi::dpl::__internal::__value_t<_Range1>;
     _Size __size;
-    template <typename _Idx, typename _AccessorSrc, typename _AccessorDst>
+
+  public:
+    __reverse_copy(_Size __size) : __size(__size) {}
+
+    template <typename _IsFull, typename _Idx>
+    void
+    __scalar_path_impl(_IsFull, const _Idx __idx, const _Range1 __rng1, _Range2 __rng2) const
+    {
+        __rng2[__idx] = __rng1[__size - __idx - 1];
+    }
+    template <typename _IsFull, typename _Idx>
+    void
+    __vector_path_impl(_IsFull __is_full, const _Idx __idx, const _Range1 __rng1, _Range2 __rng2) const
+    {
+        std::size_t __n = __size;
+        std::size_t __remaining_elements = __idx >= __n ? 0 : __n - __idx;
+        std::uint8_t __elements_to_process =
+            std::min(static_cast<std::size_t>(__base_t::__preferred_vector_size), __remaining_elements);
+        const _Idx __output_start = __size - __idx - __elements_to_process;
```

Comment:
Cant (wont this always) this underflow for non-full cases?

Reply:
I believe I've tracked down to the stride recommender that this is a `std::size_t`, at least sometimes.
Also, is it ever not a `std::size_t`?  If it is always a `std::size_t` then lets call it that rather than a template param.

Reply:
For the non-full case, `__idx < __size` should always hold which can be seen in the non-full case of `__strided_loop`. It ensures that we do not dispatch an index that is greater than or equal to the size. Actually the logic I have above this for `__remaining_elements` is unnecessary and can be removed.

Throughout this PR I think I can make all of these brick `_Idx` types just `std::size_t` as that is what gets passed through. The original implementations were templates but I think it's unnecessary.

Reply:
Ah my mistake.  I think my "Friday afternoon brain" was thinking we were subtracting the vector size rather than one shrunk by to a maximum of the elements remaining.